contracts/smart-account/modules/MultiOwnedECDSAModule.sol

```diff
@@ -0,0 +1,215 @@
+    function addOwner(address owner) external override {
+        if (_isSmartContract(owner)) revert NotEOA(owner);
+        if (owner == address(0)) revert ZeroAddressNotAllowedAsOwner();
+        if (_smartAccountOwners[owner][msg.sender])
+            revert OwnerAlreadyUsedForSmartAccount(owner, msg.sender);
+
+        _smartAccountOwners[owner][msg.sender] = true;
+        numberOfOwners[msg.sender]++;
+    }
+
+    /// @inheritdoc IMultiOwnedECDSAModule
+    function removeOwner(address owner) external override {
+        _transferOwnership(msg.sender, owner, address(0));
+        --numberOfOwners[msg.sender];
+    }
+
+    /// @inheritdoc IMultiOwnedECDSAModule
+    function isOwner(
+        address smartAccount,
+        address eoaAddress
+    ) external view override returns (bool) {
+        return _smartAccountOwners[eoaAddress][smartAccount];
+    }
+
+    /// @inheritdoc IAuthorizationModule
+    function validateUserOp(
+        UserOperation calldata userOp,
+        bytes32 userOpHash
+    ) external view virtual override returns (uint256) {
+        (bytes memory cleanEcdsaSignature, ) = abi.decode(
+            userOp.signature,
+            (bytes, address)
+        );
+        if (_verifySignature(userOpHash, cleanEcdsaSignature, userOp.sender)) {
+            return VALIDATION_SUCCESS;
+        }
+        return SIG_VALIDATION_FAILED;
+    }
+
+    /**
+     * @inheritdoc ISignatureValidator
+     * @dev Validates a signature for a message.
+     * To be called from a Smart Account.
+     * @param dataHash Exact hash of the data that was signed.
+     * @param moduleSignature Signature to be validated.
+     * @return EIP1271_MAGIC_VALUE if signature is valid, 0xffffffff otherwise.
+     */
+    function isValidSignature(
+        bytes32 dataHash,
+        bytes memory moduleSignature
+    ) public view virtual override returns (bytes4) {
+        return
+            isValidSignatureForAddress(dataHash, moduleSignature, msg.sender);
+    }
+
+    /// @inheritdoc IMultiOwnedECDSAModule
+    function isValidSignatureForAddress(
+        bytes32 dataHash,
+        bytes memory moduleSignature,
+        address smartAccount
+    ) public view virtual override returns (bytes4) {
+        if (_verifySignature(dataHash, moduleSignature, smartAccount)) {
+            return EIP1271_MAGIC_VALUE;
+        }
+        return bytes4(0xffffffff);
+    }
+
+    /// @inheritdoc IMultiOwnedECDSAModule
+    function getNumberOfOwners(
+        address smartAccount
+    ) public view returns (uint256) {
+        return numberOfOwners[smartAccount];
+    }
+
+    /**
+     * @dev Transfers ownership for smartAccount and emits an event
+     * @param newOwner Smart Account address.
+     */
+    function _transferOwnership(
+        address smartAccount,
+        address owner,
+        address newOwner
+    ) internal {
+        _smartAccountOwners[owner][smartAccount] = false;
+        _smartAccountOwners[newOwner][smartAccount] = true;
```

Comment:
could be swapped if the first value is always true and the second is always false:
similarly to this
```solidity

uint256 a = 1;
uint256 b = 2;

(a, b) = (b, a);

```

Reply:
interesting :) is it cheaper?

Reply:
if not, I would suggest not doing this as in this exact case it makes things less readable

`(_smartAccountOwners[owner][smartAccount], _smartAccountOwners[newOwner][smartAccount]) = (_smartAccountOwners[newOwner][smartAccount], _smartAccountOwners[owner][smartAccount]);`

looks bulky

Reply:
yes it's cheaper but agree that it makes it very difficult to read 😅 let's discuss when we will focus on optimization